### PR TITLE
fix(float): update position of anchored windows first

### LINKED ISSUE #14735: nvim_open_win with relative=win doesn't work if win is a float
When you use `nvim_open_win` with `relative=win` and point `win=` to a floating window, then the new float will be relative to the original non-floating window.

### Steps to reproduce using `nvim -u NORC`

```lua
local function popup(win, bg)
	win = win or vim.api.nvim_get_current_win()
	local buf = vim.api.nvim_create_buf(false, true)
	local ret = vim.api.nvim_open_win(buf, true, {
		relative = "win",
		win = win,
		col = 4,
		row = 4,
		width = 20,
		height = 10,
	})
	vim.api.nvim_win_set_option(ret, "winhl", "Normal:" .. bg .. ",EndOfBuffer:" .. bg)
	return ret
end

vim.cmd([[hi Popup1 guibg=red]])
vim.cmd([[hi Popup2 guibg=yellow]])

local p1 = popup(nil, "Popup1")
local p2 = popup(p1, "Popup2")
```

### Actual behaviour

Both the yellow and red floating window will be shown at the same place

### Expected behaviour

The second yellow window should be positioned relative to the first one, with an offset of 4 rows and 4 columns

### Other weird stuff

* When doing a vsplit of the current window, and then running the code when the second window is active, you'll see that:
  - the red popup opens in the right window
  -  the yellow popup opens in the left window (still at the wrong position)

* Probably related to #13403: `nvim_win_get_position` always returns `[0,0]` for floats
* `vim.api.nvim_win_get_number` of the floating window, returns the window number of the window where the float was created.

### LINKED ISSUE #17853: Floating window lags behind when setting `relative="win"`
### Neovim version (nvim -v)

v0.7.0-dev

### Vim (not Nvim) behaves the same?

.

### Operating system/version

NixOS 22.05

### Terminal name/version

alacritty 0.10.1

### $TERM environment variable

alacritty

### Installation

from source

### How to reproduce the issue

I'm not sure if this is a bug or I'm doing something wrong, but if a have a floating window placed relative to another floating window it seems to lag behind by 1 event when I `nvim_win_set_config`.

A demo is probably worth a thousand words, so here it is. Run `nvim --clean -u ./init.lua` with the following `init.lua`

```lua
local api = vim.api

local bufnr1 = api.nvim_create_buf(false, true)
local bufnr2 = api.nvim_create_buf(false, true)

local winid1, winid2

local opts1 = {
  relative = "cursor",
  row = 1,
  col = 0,
  width = 10,
  height = 5,
}

local opts2 = {
  relative = "win",
  win = nil,
  row = 0,
  col = opts1.width + 1,
  width = 10,
  height = 5,
}

local shift_windows = function()
  if winid1 then
    api.nvim_win_set_config(winid1, opts1)
  else
    winid1 = api.nvim_open_win(bufnr1, false, opts1)
  end

  opts2.win = winid1
  if winid2 then
    -- Closing and reopening the second window instead of setting a new config
    -- seems to fix the issue, even tough `opts2` is the exact same table in
    -- both commands.
    --
    -- api.nvim_win_hide(winid2)
    -- winid2 = api.nvim_open_win(bufnr2, false, opts2)

    api.nvim_win_set_config(winid2, opts2)
  else
    winid2 = api.nvim_open_win(bufnr2, false, opts2)
  end

  print(winid1, winid2)
end

local clear_windows = function()
  api.nvim_win_hide(winid1)
  api.nvim_win_hide(winid2)
  winid1 = nil
  winid2 = nil
end

api.nvim_create_autocmd("TextChangedI", { callback = shift_windows })
api.nvim_create_autocmd("InsertLeave", { callback = clear_windows })
```
here I'm creating two floating windows, the first one is relative to the cursor and the second one is relative to the first one. On `TextChangedI`, if they already exist they get shifted with `nvim_win_set_config`, if not they are spawned with `nvim_open_win`. Finally, they are cleared on `InsertLeave`.

I'd expect the two to move in sync, but they don't. Here's what happens

bad

as you can see, the second window is one column too close when inserting characters, and one column too far when deleting them. I think it's because it's positioning itself relative to where the first floating window used to be, not where it is now.

However, if instead of shifting the second window I close and reopen it every time, then they are in sync.

good

You can test it by uncommenting
```lua
    api.nvim_win_hide(winid2)
    winid2 = api.nvim_open_win(bufnr2, false, opts2)
```
and commenting
```lua
    -- api.nvim_win_set_config(winid2, opts2)
```
in the code above.

### Expected behavior

The two floats move together.

### Actual behavior

The second one lags behind.

### LINKED ISSUE #18925: float window position relative to another float window is not positioned correctly
### Neovim version (nvim -v)

v0.8.0-dev+404-g2b539d6fd

### Vim (not Nvim) behaves the same?

n/a

### Operating system/version

Ubuntu 20.04.4

### Terminal name/version

alacritty 0.10.0-dev

### $TERM environment variable

tmux-256color

### Installation

build from repo

### How to reproduce the issue

```lua
local api = vim.api

local parent_bufnr = api.nvim_create_buf(false, true)
local parent_winid = api.nvim_open_win(parent_bufnr, false, {
  relative = "win",
  row = 5,
  col = 10,
  height = 20,
  width = 40,
  border = "double",
  zindex = 50,
})

local child_bufnr = api.nvim_create_buf(false, true)
local child_winid = api.nvim_open_win(child_bufnr, false, {
  relative = "win",
  win = parent_winid,
  row = 1,
  col = 1,
  height = 5,
  width = 10,
  border = "rounded",
  zindex = 55,
})

vim.defer_fn(function()
  api.nvim_win_close(child_winid, true)
  api.nvim_win_close(parent_winid, true)
end, 10000)
```

### Expected behavior

Running the script would

- open the parent window positioned relative to the current window
- open the child window positioned relative to parent window

### Actual behavior

Running the script

- open the parent window positioned relative to the current window
- opens the child window positioned relative to the current window (instead of being relative to parent window)

---

#### Enter parent window after opening it

After both windows are open, if the parent window is focused by pressing `Ctrl-W` twice, the child window's position is corrected.

#### Make parent window non-focusable

If `focusable` is set to `false` for parent window:

```lua
local parent_winid = api.nvim_open_win(parent_bufnr, false, {
  focusable = false,
```

The previous case "Enter parent window after opening it" is not possible anymore.

#### Enter parent window while opening it

If the second argument to `nvim_open_win` is set to `true`:

```lua
local parent_winid = api.nvim_open_win(parent_bufnr, true, {
```

Then the parent window is focused right after opening it and the child window is positioned relative to the parent window correctly.

---

Related Issue:
- https://github.com/neovim/neovim/issues/14735

### PR DESCRIPTION
**Pull request by zeertzjq**
Fix #14735
Fix #17853
Fix #18925

### CODE REVIEW

**Comment by justinmk** [comment]
> sorting problems (probably with `pairs()`) when using PUC Lua:

FWIW we have `vim.spairs()` for in-order iteration.

File: src/nvim/window.c [hunk 1]
```diff
@@ -7614,19 +7613,50 @@ void win_get_tabwin(handle_T id, int *tabnr, int *winnr)
...
 {
+  kvec_t(win_T *) stack = KV_INITIAL_VALUE;
+
   FOR_ALL_TAB_WINDOWS(tp, wp) {
     if (wp->w_pos_changed && wp->w_grid_alloc.chars != NULL) {
       if (tp == curtab) {
+        // When a floating window is anchored to another window,
+        // update the position of its anchored window first.
+        while (true) {
+          wp->w_pos_changed = false;  // avoid looping forever in case of cycle
```
Comment (justinmk) [inline src/nvim/window.c:7624]: should we WLOG or ILOG here?

File: src/nvim/window.c [hunk 2]
```diff
@@ -7614,19 +7613,50 @@ void win_get_tabwin(handle_T id, int *tabnr, int *winnr)
...
 void win_ui_flush(bool validate)
 {
+  kvec_t(win_T *) stack = KV_INITIAL_VALUE;
+
   FOR_ALL_TAB_WINDOWS(tp, wp) {
     if (wp->w_pos_changed && wp->w_grid_alloc.chars != NULL) {
       if (tp == curtab) {
+        // When a floating window is anchored to another window,
+        // update the position of its anchored window first.
+        while (true) {
```
Comment (justinmk) [inline src/nvim/window.c:7623]: Seems worth extracting this to a function like `win_cascading_moves`

File: src/nvim/window.c [hunk 3]
```diff
@@ -7614,19 +7613,50 @@ void win_get_tabwin(handle_T id, int *tabnr, int *winnr)
...
+            break;
+          }
+          if (win->w_pos_changed && win->w_grid_alloc.chars != NULL) {
+            kv_push(stack, wp);
+            wp = win;
+            continue;
+          }
+          break;
+        }
         ui_ext_win_position(wp, validate);
```
Comment (justinmk) [inline src/nvim/window.c:7644]: `wp` here will be whatever was last assigned on line 7639, which was also added to the stack, right?
Is this line redundant?

File: src/nvim/window.c (hunk 1 above)
Comment (zeertzjq) [reply src/nvim/window.c]: Now I've changed the implementation to use recursion instead, so adding a warning for loop does add complexity.

**Pull request merged**